File: AdSwitcher-iOS/AdSwitcher/AdSwitcher/AdSwitcherBannerView.m
//
//  AdSwitcherBannerView.m
//  AdSwitcher
//
//  Created by tkyaji on 2016/08/28.
//  Copyright © 2016年 adwitcher. All rights reserved.
//

#import "AdSwitcherBannerView.h"

@implementation AdSwitcherBannerView {
    bannerAdShownHandler _bannerAdShownHandler;
    bannerAdReceived _bannerAdReceived;
    bannerAdClicked _bannerAdClicked;
    
    NSMutableDictionary<NSString *, NSObject<BannerAdAdapter> *> *_adapterCacheDict;
    NSMutableDictionary<NSString *, AdConfig *> *_adConfigDict;
    AdSelector *_adSelector;
    
    NSObject<BannerAdAdapter> *_selectedAdapter;
    AdConfig *_selectedAdConfig;
    BOOL _loading;
    BOOL _loadCancel;
    BOOL _showing;
    BOOL _autoShow;
}

- (instancetype)initWithConfigLoader:(UIViewController *)viewController configLoader:(AdSwitcherConfigLoader *)configLoader
                            category:(NSString *)category adSize:(BannerAdSize)adSize {
    return [self initWithConfigLoader:viewController configLoader:configLoader category:category adSize:adSize testMode:NO];
}

- (instancetype)initWithConfigLoader:(UIViewController *)viewController configLoader:(AdSwitcherConfigLoader *)configLoader
                            category:(NSString *)category adSize:(BannerAdSize)adSize testMode:(BOOL)testMode {
    
    CGSize cgSize = [self toCGSize:adSize];
    CGRect frame = CGRectMake(0, 0, cgSize.width, cgSize.height);
    if (self = [super initWithFrame:frame]) {
        __block AdSwitcherConfigLoader *configLoaderInBlock = configLoader;
        [configLoader addConfigLoadedHandler:^{
            AdSwitcherConfig *adSwitcherConfig = [configLoaderInBlock adSwitchConfig:category];
            [self initialize:viewController config:adSwitcherConfig adSize:adSize testMode:testMode];
        }];
    }
    
    return self;
}

- (instancetype)initWithConfig:(UIViewController *)viewController config:(AdSwitcherConfig *)adSwitcherConfig
                        adSize:(BannerAdSize)adSize {
    return [self initWithConfig:viewController config:adSwitcherConfig adSize:adSize testMode:NO];
}

- (instancetype)initWithConfig:(UIViewController *)viewController config:(AdSwitcherConfig *)adSwitcherConfig
                        adSize:(BannerAdSize)adSize testMode:(BOOL)testMode {
    
    CGSize cgSize = [self toCGSize:adSize];
    CGRect frame = CGRectMake(0, 0, cgSize.width, cgSize.height);
    if (self = [super initWithFrame:frame]) {
        [self initialize:viewController config:adSwitcherConfig adSize:adSize testMode:testMode];
    }
    
    return self;
}

- (void)load {
    [self load:false];
}

- (void)load:(BOOL)autoShow {
    _DLOG("autoShow=%d", (int)autoShow);
    if (_loading) {
        _DLOG("already started to load.");
        return;
    }
    if (_showing) {
        _DLOG("ad showing.");
        return;
    }
    _autoShow = autoShow;
    _adSelector = [[AdSelector alloc] initWithConfig:self.adSwitcherConfig];
    [self selectLoad];
}

- (void)show {
    if (![self isLoaded]) {
        _DLOG("also not loaded.");
        return;
    }
    if (_showing) {
        _DLOG("already show.");
        return;
    }
    [_selectedAdapter bannerAdShow:self];
    _showing = YES;
}

- (void)hide {
    _DLOG();
    if (_showing) {
        [_selectedAdapter bannerAdHide];
        _selectedAdapter = nil;
        _selectedAdConfig = nil;
        _showing = NO;
        
    }
    if (_loading) {
        _loadCancel = YES;
    }
}

- (void)switchAd {
    _DLOG();
    [self hide];
    [self load];
}

- (BOOL)isLoaded {
    return _selectedAdapter;
}

- (CGSize)getSize {
    return [self toCGSize:self.adSize];
}

- (void)setAdReceivedHandler:(void (^)(AdConfig *config, BOOL result))handler {
    bannerAdReceived handlerWrapper = ^(AdConfig *config, BOOL result) {
        dispatch_async(dispatch_get_main_queue(), ^{
            handler(config, result);
        });
    };
    _bannerAdReceived = handlerWrapper;
}

- (void)setAdShownHandler:(void (^)(AdConfig *config))handler {
    bannerAdShownHandler handlerWrapper = ^(AdConfig *config) {
        dispatch_async(dispatch_get_main_queue(), ^{
            handler(config);
        });
    };
    _bannerAdShownHandler = handlerWrapper;
}

- (void)setAdClickedHandler:(void (^)(AdConfig *config))handler {
    bannerAdClicked handlerWrapper = ^(AdConfig *config) {
        dispatch_async(dispatch_get_main_queue(), ^{
            handler(config);
        });
    };
    _bannerAdClicked = handlerWrapper;
}


#pragma - BannerAdDelegate

- (void)bannerAdReceived:(NSObject<AdAdapter> *)adAdapter result:(BOOL)result {
    NSString *className = NSStringFromClass(adAdapter.class);
    _DLOG(@"%@, result=%d", className, result);
    
    if (!_selectedAdConfig || ![_selectedAdConfig.className isEqualToString:className]) {
        _DLOG(@"class name is invalid. selectedClass=%@", _selectedAdConfig.className);
        return;
    }
    
    _loading = false;
    
    if (!_selectedAdapter && result) {
        _selectedAdapter = (NSObject<BannerAdAdapter> *)adAdapter;
    }
    
    if (_bannerAdReceived) {
        _bannerAdReceived([_adConfigDict objectForKey:className], result);
    }
    
    if (result) {
        if (_autoShow) {
            [self show];
        }
    } else {
        [self selectLoad];
    }
}

- (void)bannerAdShown:(NSObject<AdAdapter> *)adAdapter {
    NSString *className = NSStringFromClass(adAdapter.class);
    _DLOG(@"%@", className);
    if (_bannerAdShownHandler) {
        _bannerAdShownHandler([_adConfigDict objectForKey:className]);
    }
}

- (void)bannerAdClicked:(NSObject<AdAdapter> *)adAdapter {
    NSString *className = NSStringFromClass(adAdapter.class);
    _DLOG(@"%@", className);
    if (_bannerAdClicked) {
        _bannerAdClicked([_adConfigDict objectForKey:className]);
    }
}


#pragma - UIView

- (void)removeFromSuperview {
    if (_selectedAdapter) {
        [_selectedAdapter bannerAdHide];
    }
    [super removeFromSuperview];
}


#pragma - private methods

- (void)initialize:(UIViewController *)viewController config:(AdSwitcherConfig *)adSwitcherConfig
            adSize:(BannerAdSize)adSize testMode:(BOOL)testMode {
    
    _DLOG("testMode=%d, adSize=%d, switchType=%d", testMode, (int)adSize, (int)adSwitcherConfig.switchType);

    self.viewController = viewController;
    self.adSwitcherConfig = adSwitcherConfig;
    self.testMode = testMode;
    self.adSize = adSize;
    
    _adapterCacheDict = [NSMutableDictionary<NSString *, NSObject<BannerAdAdapter> *> new];
    
    _adConfigDict = [NSMutableDictionary new];
    for (AdConfig *adConfig in adSwitcherConfig.adConfigArr) {
        [_adConfigDict setObject:adConfig forKey:adConfig.className];
    }
}

- (void)selectLoad {
    if (_loadCancel) {
        _loading = NO;
        _loadCancel = NO;
        return;
    }
    _loading = YES;
    
    NSObject<BannerAdAdapter> *bannerAdAdapter = nil;
    while (!bannerAdAdapter) {
        _selectedAdConfig = [_adSelector selectAd];
        if (!_selectedAdConfig) {
            break;
        }
        
        [self initAdapter:_selectedAdConfig];
        bannerAdAdapter = [_adapterCacheDict objectForKey:_selectedAdConfig.className];
    }
    
    if (bannerAdAdapter) {
        _DLOG("%@", NSStringFromClass(bannerAdAdapter.class));
        [bannerAdAdapter bannerAdLoad];
        
    } else {
        _DLOG("It will not be able to display all.");
        dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC);
        dispatch_after(time, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            dispatch_async(dispatch_get_main_queue(), ^{
                _DLOG("retry load.");
                _adSelector = [[AdSelector alloc] initWithConfig:self.adSwitcherConfig];
                [self selectLoad];
            });
        });
    }
}

- (void)initAdapter:(AdConfig *)adConfig {
    
    if ([_adapterCacheDict objectForKey:adConfig.className]) {
        return;
    }
    
    Class adAdapterClass = NSClassFromString(adConfig.className);
    if (!adAdapterClass) {
        _DLOG("%@ class is not found.", adConfig.className);
        return;
    }
    
    NSObject<AdAdapter> *adAdapter = [adAdapterClass new];
    
    if (![adAdapter conformsToProtocol:@protocol(BannerAdAdapter)]) {
        _DLOG(@"%@ class is not conform BannerAdAdapter.", adConfig.className);
        return;
    }
    
    NSObject<BannerAdAdapter> *bannerAdAdapter = (NSObject<BannerAdAdapter> *)adAdapter;
    
    @try {
        [bannerAdAdapter setValue:self forKey:@"bannerAdDelegate"];
        
    } @catch (NSException *exception) {
        _DLOG(@"bannerAdDelegate property is not found. : %@", adConfig.className);
        return;
    }
    
    [bannerAdAdapter bannerAdInitialize:self.viewController parameters:adConfig.parameters testMode:self.testMode adSize:self.adSize];
    
    [_adapterCacheDict setObject:bannerAdAdapter forKey:adConfig.className];
}

- (CGSize)toCGSize:(BannerAdSize)adSize {
    switch (adSize) {
        case BannerAdSize_320x50:
            return CGSizeMake(320, 50);
            
        case BannerAdSize_320x100:
            return CGSizeMake(320, 100);
            
        case BannerAdSize_300x250:
            return CGSizeMake(300, 250);
    }
    return CGSizeMake(320, 50);
}

@end

File: AdSwitcher-Android/adswitcher/src/main/java/net/adswitcher/AdSwitcherBannerView.java
package net.adswitcher;

import android.app.Activity;
import android.os.Build;
import android.os.Handler;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Display;
import android.widget.FrameLayout;

import net.adswitcher.adapter.AdAdapter;
import net.adswitcher.adapter.BannerAdAdapter;
import net.adswitcher.adapter.BannerAdListener;
import net.adswitcher.adapter.BannerAdSize;
import net.adswitcher.config.AdConfig;
import net.adswitcher.config.AdSwitcherConfig;
import net.adswitcher.config.AdSwitcherConfigLoader;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by tkyaji on 2016/09/04.
 */
public class AdSwitcherBannerView extends FrameLayout implements BannerAdListener {

    private static final String TAG = "AdSwitcherBannerView";

    private AdReceivedListener adReceivedListener;
    private AdShownListener adShownListener;
    private AdClickedListener adClickedListener;

    private Activity activity;
    private AdSwitcherConfig adSwitcherConfig;
    private boolean testMode;
    private BannerAdSize adSize;

    private BannerAdAdapter selectedAdapter;
    private AdConfig selectedAdConfig;
    private AdSelector adSelector;
    private Map<String, BannerAdAdapter> adapterCacheMap;
    private Map<String, AdConfig> adConfigMap;
    private boolean loading;
    private boolean loadCancel;
    private boolean showing;
    private boolean autoShow;


    public AdSwitcherBannerView(final Activity activity, final AdSwitcherConfigLoader configLoader,
                                final String category, final BannerAdSize adSize) {
        this(activity, configLoader, category, adSize, false);
    }

    public AdSwitcherBannerView(final Activity activity, final AdSwitcherConfigLoader configLoader,
                                final String category, final BannerAdSize adSize, final boolean testMode) {
        super(activity);
        this.activity = activity;
        this.adSize = adSize;
        this.testMode = testMode;

        configLoader.addConfigLoadedHandler(new AdSwitcherConfigLoader.ConfigLoadHandler() {
            @Override
            public void onLoaded() {
                final AdSwitcherConfig adSwitcherConfig = configLoader.getAdSwitcherConfig(category);
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        AdSwitcherBannerView.this.initialize(adSwitcherConfig);
                    }
                });
            }
        });
    }

    public AdSwitcherBannerView(final Activity activity, final AdSwitcherConfig adSwitcherConfig,
                                final BannerAdSize adSize) {
        this(activity, adSwitcherConfig, adSize, false);
    }

    public AdSwitcherBannerView(final Activity activity, final AdSwitcherConfig adSwitcherConfig,
                                final BannerAdSize adSize, final boolean testMode) {
        super(activity);
        this.activity = activity;
        this.testMode = testMode;
        this.adSize = adSize;

        this.initialize(adSwitcherConfig);
    }

    public void load() {
        this.load(false);
    }

    public void load(boolean autoShow) {
        Log.d(TAG, "load : autoShow=" + autoShow);
        if (this.loading) {
            Log.d(TAG, "already started to load.");
            return;
        }
        if (this.showing) {
            Log.d(TAG, "ad showing.");
            return;
        }
        this.autoShow = autoShow;
        this.adSelector = new AdSelector(this.adSwitcherConfig);
        this.selectLoad();
    }

    public void show() {
        if (!this.isLoaded()) {
            Log.d(TAG, "also not loaded.");
            return;
        }
        if (this.showing) {
            Log.d(TAG, "already shown.");
            return;
        }
        this.selectedAdapter.bannerAdShow(this);
        this.showing = true;
    }

    public void hide() {
        Log.d(TAG, "hide");
        if (this.showing) {
            this.selectedAdapter.bannerAdHide();
            this.selectedAdapter = null;
            this.selectedAdConfig = null;
            this.showing = false;
        }
        if (this.loading) {
            this.loadCancel = true;
        }
    }

    public void switchAd() {
        Log.d(TAG, "switchAd");
        this.hide();
        this.load(true);
    }

    public boolean isLoaded() {
        return (this.selectedAdapter != null);
    }

    public float getDpiWidth() {
        switch (this.adSize) {
            case SIZE_320X50:
            case SIZE_320X100:
                return 320;
            case SIZE_300X250:
                return 300;
        }
        return 0;
    }

    public float getDpiHeight() {
        switch (this.adSize) {
            case SIZE_320X50:
                return 50;
            case SIZE_320X100:
                return 100;
            case SIZE_300X250:
                return 250;
        }
        return 0;
    }

    public float getPxWidth() {
        float dpiW = this.getDpiWidth();
        DisplayMetrics metrics = this.getMetrics();
        return dpiW * metrics.density;
    }

    public float getPxHeight() {
        float dpiH = this.getDpiHeight();
        DisplayMetrics metrics = this.getMetrics();
        return dpiH * metrics.density;
    }

    private DisplayMetrics getMetrics() {
        Display display = this.activity.getWindowManager().getDefaultDisplay();
        DisplayMetrics metrics = new DisplayMetrics();
        if (Build.VERSION.SDK_INT >= 17) {
            display.getRealMetrics(metrics);
        } else {
            display.getMetrics(metrics);
        }
        return metrics;
    }

    public void setAdReceivedListener(final AdReceivedListener listener) {
        this.adReceivedListener = new AdReceivedListener() {
            @Override
            public void onAdReceived(final AdConfig config, final boolean result) {
                AdSwitcherBannerView.this.activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        listener.onAdReceived(config, result);
                    }
                });
            }
        };
    }

    public void setAdShownListener(final AdShownListener listener) {
        this.adShownListener = new AdShownListener() {
            @Override
            public void onAdShown(final AdConfig config) {
                AdSwitcherBannerView.this.activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        listener.onAdShown(config);
                    }
                });
            }
        };
    }

    public void setAdClickedListener(final AdClickedListener listener) {
        this.adClickedListener = new AdClickedListener() {
            @Override
            public void onAdClicked(final AdConfig config) {
                AdSwitcherBannerView.this.activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        listener.onAdClicked(config);
                    }
                });
            }
        };
    }


    private void initialize(AdSwitcherConfig adSwitcherConfig) {
        Log.d(TAG, "testMode=" + testMode + ", adSize=" + adSize);

        this.adSwitcherConfig = adSwitcherConfig;

        this.adapterCacheMap = new HashMap<>();
        this.adConfigMap = new HashMap<>();

        if (this.adSwitcherConfig != null) {
            for (AdConfig config : this.adSwitcherConfig.adConfigList) {
                this.adConfigMap.put(config.className, config);
            }
        }
    }

    private void selectLoad() {
        if (this.loadCancel) {
            this.loading = false;
            this.loadCancel = false;
            return;
        }
        this.loading = true;

        BannerAdAdapter bannerAdAdapter = null;
        while (bannerAdAdapter == null) {
            this.selectedAdConfig = this.adSelector.selectAd();
            if (this.selectedAdConfig == null) {
                break;
            }

            this.initAdapter(this.selectedAdConfig);
            bannerAdAdapter = this.adapterCacheMap.get(this.selectedAdConfig.className);
        }

        if (bannerAdAdapter != null) {
            Log.d(TAG, bannerAdAdapter.getClass().getName());
            final BannerAdAdapter _bannerAdAdapter = bannerAdAdapter;
            AdSwitcherBannerView.this.activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    _bannerAdAdapter.bannerAdLoad();
                }
            });

        } else {
            Log.w(TAG, "It will not be able to display all.");

            (new Handler()).postDelayed(new Runnable() {
                @Override
                public void run() {
                    AdSwitcherBannerView.this.activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Log.d(TAG, "load retry");
                            AdSwitcherBannerView.this.adSelector = new AdSelector(AdSwitcherBannerView.this.adSwitcherConfig);
                            AdSwitcherBannerView.this.selectLoad();
                        }
                    });
                }
            }, 5000);
        }
    }

    private void initAdapter(final AdConfig adConfig) {
        if (this.adapterCacheMap.containsKey(adConfig.className)) {
            return;
        }

        try {
            Class<?> adAdapterClass = Class.forName(adConfig.className);
            final BannerAdAdapter bannerAdAdapter = (BannerAdAdapter) adAdapterClass.newInstance();

            this.activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    bannerAdAdapter.bannerAdInitialize(AdSwitcherBannerView.this.activity,
                            AdSwitcherBannerView.this, adConfig.parameters,
                            AdSwitcherBannerView.this.testMode, AdSwitcherBannerView.this.adSize);
                }
            });
            this.adapterCacheMap.put(adConfig.className, bannerAdAdapter);

        } catch (ClassNotFoundException ex) {
            Log.w(TAG, adConfig.className + " class is not found.");
            return;

        } catch (Throwable e) {
            Log.w(TAG, adConfig.className + " class is not conform BannerAdAdapter.");
            return;
        }
    }


    @Override
    public void bannerAdReceived(AdAdapter adAdapter, boolean result) {
        String className = adAdapter.getClass().getName();
        Log.d(TAG, "bannerAdReceived : " + className + ", result=" + result);

        if (this.selectedAdConfig == null) {
            Log.d(TAG, "class name is invalid.(null)");
            return;
        } else if (!this.selectedAdConfig.className.equals(className)) {
            Log.d(TAG, "class name is invalid. selectedClass=" + this.selectedAdConfig.className);
            return;
        }

        this.loading = false;

        if (this.selectedAdapter == null && result) {
            this.selectedAdapter = (BannerAdAdapter) adAdapter;
        }

        if (this.adReceivedListener != null) {
            this.adReceivedListener.onAdReceived(this.adConfigMap.get(className), result);
        }

        if (this.showing) {
            return;
        }

        if (result) {
            if (this.autoShow) {
                AdSwitcherBannerView.this.activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        AdSwitcherBannerView.this.show();
                    }
                });
            }
        } else {
            this.selectLoad();
        }
    }

    @Override
    public void bannerAdShown(AdAdapter adAdapter) {
        String className = adAdapter.getClass().getName();
        Log.d(TAG, "bannerAdShown : " + className);

        if (this.adShownListener != null) {
            this.adShownListener.onAdShown(this.adConfigMap.get(className));
        }
    }

    @Override
    public void bannerAdClicked(AdAdapter adAdapter) {
        String className = adAdapter.getClass().getName();
        Log.d(TAG, "bannerAdClicked : " + className);

        if (this.adClickedListener != null) {
            this.adClickedListener.onAdClicked(this.adConfigMap.get(className));
        }
    }


    public interface AdReceivedListener {
        void onAdReceived(AdConfig config, boolean result);
    }

    public interface AdShownListener {
        void onAdShown(AdConfig config);
    }

    public interface AdClickedListener {
        void onAdClicked(AdConfig config);
    }

}
